IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -443,44 +567,135 @@ - (MSIDCredentialCacheItem *)tokenWithKey:(__unused MSIDCacheKey *)key
         }
     }
-    return tokenList;
 }
 
 // Remove one or more credentials from the keychain that match the key (see credentialItem:matchesKey).
-- (BOOL)removeTokensWithKey:(__unused MSIDCacheKey *)key
-                    context:(__unused id<MSIDRequestContext>)context
-                      error:(__unused NSError **)error
+- (BOOL)removeTokensWithKey:(MSIDCacheKey *)key context:(id<MSIDRequestContext>)context error:(NSError *__autoreleasing *)error
 {
-    return [self removeItemsWithKey:key context:context error:error];
+    MSID_TRACE;
+    
+    if (!key || !(key.account || key.service))
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Key is nil or one of the key attributes account or service is nil.");
+        [self createError:@"Key is nil or one of the key attributes account or service is nil."
+                   domain:MSIDErrorDomain errorCode:MSIDErrorInvalidDeveloperParameter error:error context:context];
+        return NO;
+    }
+    
+    MSIDMacCredentialStorageItem *storageItem = [self storageItemWithKey:key serializer:self.serializer context:context error:error];
+    
+    if (storageItem)
+    {
+        NSString *credentialKey = [NSString stringWithFormat:@"%@%@%@", key.account, keyDelimiter, key.service];
+        [storageItem removeStoredCredentialForKey:credentialKey];
+        return [self saveStorageItem:storageItem key:key serializer:self.serializer context:context error:error];
+    }
+    
+    return YES;
+}
+
+- (MSIDMacCredentialStorageItem *)storageItemWithKey:(MSIDCacheKey *)key
+                                          serializer:(id<MSIDCacheItemSerializing>)serializer
+                                             context:(id<MSIDRequestContext>)context
+                                               error:(NSError **)error
+{
+    MSID_TRACE;
+    MSIDMacCredentialStorageItem *storageItem = nil;
+    NSMutableDictionary *query = [self.defaultCacheQuery mutableCopy];
+    
+    if (key.isShared)
+    {
+        // Shared item attribute: <keychainGroup>
+        query[(id)kSecAttrAccount] = self.keychainGroup;
+    }
+    else
+    {
+        // Secret item attribute: <keychainGroup>-<app_bundle_id>
+        query[(id)kSecAttrAccount] = [NSString stringWithFormat:@"%@-%@", self.keychainGroup, [[NSBundle mainBundle] bundleIdentifier]];
+    }
+    
+    query[(id)kSecAttrService] = s_defaultKeychainLabel;
+    query[(id)kSecReturnAttributes] = (__bridge id)kCFBooleanTrue;
+    query[(id)kSecReturnData] = (__bridge id)kCFBooleanTrue;
+    
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"Trying to find keychain items...");
+    
+    __block CFDictionaryRef result = nil;
+    __block OSStatus status;
+    
+    dispatch_sync(s_synchronizationQueue, ^{
+        status = SecItemCopyMatching((CFDictionaryRef)query, (CFTypeRef *)&result);
+    });
+    
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, nil, @"Keychain find status: %d.", (int)status);
+    
+    if (status == errSecSuccess)
```

Comment:
log if not found
log and return error if error encountered

Reply:
Added.